### DISCUSSION TITLE
multiple themes in single jar file

### DISCUSSION THREAD

**Discussion by tonythomas1** [keycloakify/keycloakify#558] [General]
I wanted to try and implement multiple themes ( each theme for different client ) , and using the client id use it to route the css differently for each page. was wondering if this has been done before or even if it would be possible. Thanks!

**Comment by garronej** [comment]
Hello @tonythomas1,  

Yes, this is something that is fully supported! 

https://docs.keycloakify.dev/build-options#themename

**Reply by tonythomas1** [reply]
I think I should have explained it better but the situation is that I have a realm and multiple realm roles within it, and on applying a single theme for the realm, I want different themes being applied to the realm roles, is this possible ?

**Reply by garronej** [reply]
Please explain exactly what you want to achieve and I'll try to point you to the right direction

**Reply by tonythomas1** [reply]
From what i understand keycloakify supports bundling themes in the jar file and you can choose and apply these themes to your respective realm.what i wanted to know was would it possible to apply a theme to a realm which routes the css and images based on client id.
 i hope my explanation makes sense but basically i want one theme with logic to route css based on a parameter that can differentiate between clients and that parameter so far looks like client id to me but i am not sure if this is possible.

**Reply by garronej** [reply]
Ok,  
Yes you can select login theme on a per client basis.  
I think you went under realm settings -> theme and saw that you could select theme here for all the realm and assumed that this was the only way to apply the theme.  
But actually you can select the login theme in the client configuration:  


https://github.com/keycloakify/keycloakify/assets/6702424/cfb38df8-e1ff-4e8b-a5be-1f971dea9e5d

**Reply by tonythomas1** [reply]
oh wow, don't know how I didn't see this but thanks a lot!, gonna play with this and see how it turns out. thanks for the quick response and help :)

**Reply by denyFh** [reply]
Hello @tonythomas1, May I ask how can you create multiple theme and bundling them inside a single jar file? I was stuck on doing something similar for my docker keycloak

**Reply by garronej** [reply]
Hello @denyFh,  
Have you checked out this documentation section: https://docs.keycloakify.dev/theme-variants  
Seems to be specifically what you are looking for.  

Now at a higher level, I'll add a feature for bundling extra jars files since pepoles seems to need it but keep in mind that jar files are just .zip(ish) files at the end of the day you can extract them, merge the files then re-archive the files into a single .jar file.

**Reply by tonythomas1** [reply]
> Hello @tonythomas1, May I ask how can you create multiple theme and bundling them inside a single jar file? I was stuck on doing something similar for my docker keycloak

yes i use the themeName approach https://docs.keycloakify.dev/theme-variants and then conditionally render css based on the theme name, works well

**Reply by denyFh** [reply]
Thanks a lot! @garronej @tonythomas1 at first, I was doing experiment on my Dockerfile to create a multiple theme based on each folder with different theme styles since my assumption of the .jar file created can be individually added into the keycloak /providers which turns failed because the .jar file name is the same so it overwrites each other. I'll try again with the docs guidance now!
